src/poll.rs

```diff
@@ -393,115 +717,202 @@ impl fmt::Debug for Registration {
-    fn queue_for_processing(&self) -> bool {
-        // `Release` ensures that the `events` mutation is visible if this
-        // mutation is visible.
+        let mut next_token_pos = curr_token_pos;
+
+        // If the `update` call is changing the token, then compute the next
+        // available token slot and write the token there.
         //
-        // `Acquire` ensures that a change to `head_readiness` made in the
-        // poll thread is visible if `queued` has been reset to zero.
-        let prev = self.node().queued.compare_and_swap(0, NODE_QUEUED_FLAG, Ordering::AcqRel);
-
-        // If the queued flag was not initially set, then the current thread
-        // is assigned the responsibility of enqueuing the node for processing.
-        if prev == 0 {
-            self.queue.prepend_readiness_node(self.node.clone())
-        } else {
-            false
+        // Note that this computation is happening *outside* of the
+        // compare-and-swap loop. The update lock ensures that only a single
+        // thread could be mutating the write_token_position, so the
+        // `next_token_pos` will never need to be recomputed even if
+        // `token_read_pos` concurrently changes. This is because
+        // `token_read_pos` can ONLY concurrently change to the current value of
+        // `token_write_pos`, so `next_token_pos` will always remain valid.
+        if token != curr_token {
+            next_token_pos = state.next_token_pos();
+
+            // Update the token
+            match next_token_pos {
+                0 => unsafe { *self.token_0.get() = token },
+                1 => unsafe { *self.token_1.get() = token },
+                2 => unsafe { *self.token_2.get() = token },
+                _ => unreachable!(),
+            }
         }
-    }
 
-    fn node(&self) -> &ReadinessNode {
-        self.node.as_ref().unwrap()
-    }
+        // Now enter the compare-and-swap loop
+        loop {
+            next = state;
 
-    fn registration_data_mut(&self, readiness_queue: &ReadinessQueue) -> io::Result<&mut RegistrationData> {
-        // `&Poll` is passed in here in order to ensure that this function is
-        // only called from the thread that owns the `Poll` value. This is
-        // required because the function will mutate variables that are read
-        // from a call to `Poll::poll`.
+            // The node is only dropped once all `Registration` handles are
+            // dropped. Only `Registration` can call `update`.
+            debug_assert!(!state.is_dropped());
 
-        if !self.queue.identical(readiness_queue) {
-            return Err(io::Error::new(io::ErrorKind::Other, "registration registered with another instance of Poll"));
+            // Update the write token position, this will also release the token
+            // to Poll::poll.
+            if curr_token != token {
+                next.set_token_write_pos(next_token_pos);
```

Comment:
I think I'm missing something, but could this unconditionally happen? We only change `next_token_pos` in the case the tokens are different, so I'd naively expect this to be a noop if the tokens were equal (e.g. could happen unconditionally)